src/metadata/metadata.rs

```diff
@@ -0,0 +1,207 @@
+use chrono::{Datelike, NaiveDate};
+use regex::Regex;
+use titlecase::titlecase;
+
+use super::join::Joinable;
+use super::parse;
+use crate::config::SHORT_TO_LONG_STATUS;
+use crate::line::Line;
+use crate::spec::Spec;
+
+pub type Date = NaiveDate;
+
+#[derive(Debug, Default, Clone)]
+pub struct MetadataManager {
```

Comment:
`MetadataManager` seems kind of a weird name for something that is just storing data, it doesn't "manage" much, right? I'd just call this `Metadata` or `SpecMetadata` if you want.

So it seems this could be simpler unless I'm missing something... What this is doing is keeping all the metadata of the document, from three different sources (the "baseline", which is just the date), the document, and the cli, in that order.

It seems this could just be:

```rust
pub struct Metadata {
    pub abs: Vec<String>,
    pub canonical_url: Option<String>,
    pub date: Date,
    pub ed: Option<String>,
    pub editors: Vec<String>,
    pub group: Option<String>,
    pub level: Option<String>,
    pub shortname: Option<String>,
    ...
}

impl Metadata {
    pub fn new() -> Self {
        Self {
            date: parse_date("now"),
            // Options initialized to `None`, vectors to `vec![]`.
        }
    }

    fn add_data(&mut self, ..) {
        // roughly as-is
    }

    fn join(&mut self, other: Self) {
        self.abs.extend(other.abs.into_iter());
        if let Some(canonical_url) = other.canonical_url { // Maybe a macro for this would be on point.
            self.canonical_url = canonical_url;
        }
    }
```

Then you'd just need to do something like:

```rust
let (lines, mut metadata) = parse_metadata(&lines, &mut metadata);
metadata.join(cli_metadata);
```

or such. That way you don't need the `Joinable` trait and such, unless I'm missing something.

Reply:
Considering there're about 80 keys in the metadata, I add `Joinable` to keep the code simple. Do you think that's a good idea when there are so many fields in the `Metadata` struct?

Reply:
What part of it in particular?

I don't think there's a reason to have `Option<Vec<>>` if we're just going to extend the `Vec`.

Even with the Joinable trait, there's a place where we need to enumerate them all anyway (when merging metadata), right? If so, I don't think it brings much value. But even if you want to keep it, we should probably make it simpler. There's no reason the `Joinable` trait needs to be generic afaict, right?

```rust
trait Joinable {
    fn join(&mut self, other: &Self);
}

impl<T> Joinable for Option<T>
where
    T: Clone,
{
    fn join(&mut self, other: &Self) {
        if let Some(v) = other {
            *self = Some(v.clone());
        }
    }
}

impl<T> Joinable for Vec<T>
where
    T: Clone,
{
    fn join(&mut self, other: &Self) {
        self.extend(other.iter().cloned());
    }
}

impl Joinable for Metadata {
    fn join(&mut self, other: &Self) {
        // ...
    }
}
```

would do for our purposes. That opens the future to potentially `#[derive]` it in the future with a proc macro, so that metadata would eventually look like:

```rust
#[derive(Joinable)]
struct Metadata {
    pub has_metadata: bool,
    pub date: Option<Date>,
    editors: Vec<String>,
    // or such
}
```

But I'm still on the fence on whether that's worth the complexity. Anyhow, it's probably fine to keep, but then it should be simpler (as described above) unless I'm missing something.